### PR TITLE
Configuring error codes in CloudFront

### PR DESCRIPTION
**Pull request by giorgiosironi**
Pages are hosted into a publicly readable bucket, CloudFront allows to:
- point all requests that match `???.html` to this bucket
- specify a fixed set of error codes to be mapped to these paths.

### CODE REVIEW

File: projects/elife.yaml [hunk 1]
```diff
@@ -223,6 +238,18 @@ journal:
...
+                subdomains:
+                    - "{instance}--cdn-journal"
+                    - journal # in the future 'beta' or... ''?
+                errors: 
+                    domain: prod-elife-error-pages.s3-website-us-east-1.amazonaws.com
+                    pattern: "???.html"
+                    codes:
+                        - 404
+                        # not supported
+                        # - 410
```
Comment (giorgiosironi) [inline projects/elife.yaml]: @thewilkybarkid we won't have 410 according to the docs

File: projects/elife.yaml [hunk 2]
```diff
@@ -205,15 +206,29 @@ journal:
...
+                        - 404
+                        # not supported
+                        # - 410 
+                        - 503
+        continuumtest:
+            cloudfront:
+                subdomains:
+                    - "{instance}--cdn-journal"
+                errors: 
+                    domain: prod-elife-error-pages.s3-website-us-east-1.amazonaws.com
```
Comment (giorgiosironi) [inline projects/elife.yaml]: doesn't have its own `error-pages--continuumtest` instance, maybe it should

File: src/buildercore/trop.py [hunk 3]
```diff
@@ -546,6 +546,36 @@ def render_cloudfront(context, template, origin_hostname):
             SslSupportMethod='sni-only'
         )
     }
+    if context['cloudfront']['errors']:
+        props['Origins'].append(cloudfront.Origin(
+            DomainName=context['cloudfront']['errors']['domain'],
+            # TODO: constant
+            Id='errors',
```
Comment (giorgiosironi) [inline src/buildercore/trop.py]: yeah, I need to extract that constant

File: projects/elife.yaml (hunk 1 above)
Comment (thewilkybarkid) [inline projects/elife.yaml]: Makes sense that CloudFront wouldn't produce it.

File: src/buildercore/trop.py [hunk 4]
```diff
@@ -546,6 +546,36 @@ def render_cloudfront(context, template, origin_hostname):
...
+            cloudfront.CacheBehavior(
+                TargetOriginId='errors',
+                ForwardedValues=cloudfront.ForwardedValues(
+                    QueryString=False
+                ),
+                PathPattern=context['cloudfront']['errors']['pattern'],
+                ViewerProtocolPolicy='allow-all',
+            ),
+        ]
+        all_codes = [400, 403, 404, 405, 414, 416, 500, 501, 502, 503, 504]
```
Comment (thewilkybarkid) [inline src/buildercore/trop.py]: Reading the CloudFront docs again, does this mean that a `404 Not Found` from Journal will result in `/404.html` being returned instead? We _only_ want these custom error pages for CloudFront-produced errors.

Comment (giorgiosironi) [inline src/buildercore/trop.py]: Then we should only configure `502` according to the docs: http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/troubleshooting-response-errors.html
However, I've seen forum reports for 503 and 504, so maybe all three?

Comment (thewilkybarkid) [inline src/buildercore/trop.py]: Hmm, probably not `503` and `504`, Journal can produce those.

`414` probably should be too.

(Again, something that should be easier with Fastly.)

**Pull request merged**